bundles/org.openhab.binding.netatmo/src/main/java/org/openhab/binding/netatmo/internal/api/ApiBridge.java

```diff
@@ -0,0 +1,233 @@
+    private final Gson gson;
+
+    private NetatmoBindingConfiguration configuration = new NetatmoBindingConfiguration();
+    private Map<Class<? extends RestManager>, RestManager> managers = new HashMap<>();
+    private ConnectionStatus connectionStatus = ConnectionStatus.UNKNOWN;
+    private List<Scope> grantedScopes = List.of();
+
+    private @Nullable ScheduledFuture<?> connectJob;
+
+    @Activate
+    public ApiBridge(@Reference HttpClientFactory httpClientFactory, @Reference NetatmoGson netatmoGson,
+            ComponentContext componentContext) {
+        this.httpClient = httpClientFactory.getCommonHttpClient();
+        this.httpHeaders.put(HttpHeader.CONTENT_TYPE, "application/x-www-form-urlencoded;charset=UTF-8");
+        this.gson = netatmoGson.getGson();
+        this.connectApi = new AuthenticationApi(this, configuration, scheduler);
+
+        openConnection(BindingUtils.ComponentContextToMap(componentContext));
+    }
+
+    @Modified
+    public void openConnection(@Nullable Map<String, Object> config) {
+        try {
+            configuration.update(
+                    config != null ? new Configuration(config).as(NetatmoBindingConfiguration.class) : configuration);
+            logger.debug("Updated binding configuration to {}", configuration);
+            connectApi.authenticate();
+            setConnectionStatus(ConnectionStatus.SUCCESS);
+        } catch (NetatmoException e) {
+            prepareReconnection(e);
+        }
+    }
+
+    @Deactivate
+    public void dispose() {
+        connectApi.dispose();
+        freeConnectJob();
+    }
+
+    private void freeConnectJob() {
+        ScheduledFuture<?> job = connectJob;
+        if (job != null) {
+            job.cancel(true);
+        }
+        connectJob = null;
+    }
+
+    private void prepareReconnection(NetatmoException e) {
+        setConnectionStatus(ConnectionStatus.Failed("Will retry to connect Netatmo API, this call failed : %s", e));
+        onAccessTokenResponse(null, List.of());
+        freeConnectJob();
+        connectJob = scheduler.schedule(() -> openConnection(null), configuration.reconnectInterval, TimeUnit.SECONDS);
+    }
+
+    @SuppressWarnings("unchecked")
+    public <T extends RestManager> T getRestManager(Class<T> typeOfRest) {
+        if (!managers.containsKey(typeOfRest)) {
+            try {
+                Constructor<T> constructor = typeOfRest.getConstructor(ApiBridge.class);
+                T tentative = constructor.newInstance(this);
+                if (!grantedScopes.containsAll(tentative.getRequiredScopes())) {
+                    throw new NetatmoException("Required scopes missing to access : " + typeOfRest);
+                }
+                managers.put(typeOfRest, tentative);
+            } catch (SecurityException | ReflectiveOperationException | NetatmoException e) {
+                logger.error("Error invoking RestManager constructor for class {} : {}", typeOfRest, e.getMessage());
```

Comment:
Well, I logged this as an error because it should only happen if making a mistake in development, this should never arrive in production. Your thoughts ?

Reply:
Code guidelines: https://www.openhab.org/docs/developer/guidelines.html#f-logging

> Logging levels should focus on the system itself and describe its state. As every bundle is only one out of many, logging should be done very scarce. It should be up to the user to increase the logging level for specific bundles, packages or classes if necessary. This means in detail:
> 
>     error logging should only be used
>         to inform the user that something is tremendously wrong in the setup, the system cannot function normally anymore, and there is a need for immediate action.
>         in case some code fails irrecoverably and the user should report it as a severe bug.
> 
> warn logging should only be used
> 
>     to inform the user that something seems to be wrong in the overall setup, but the system can nonetheless function as normal,
>     in recoverable situations when a section of code that is not accessed under normal operating conditions is reached.

IMHO, you should use WARN level.

Reply:
Ok, changed.